Refactor SAGE

jzelenjak opened the pull request:
## Description

Applies refactoring to SAGE from the beginning until the end.

In particular, this PR:
* Removes unnecessary comments (unused code that is commented out) and unused code blocks
* Changes some variable names to improve readability
* Simplifies parts of the code that can be simplified
* Extracts some parts of long functions into separate functions (e.g `_plot_alert_volume_per_episode`, `_create_episode`, `plot_alert_filtering` etc.)
* Addresses PEP warnings

## Future work

The following issues have been created as a result of this PR:
* #24
* #25 
* #26 
* #27 
* #28
* #29 
* #30

jzelenjak committed:
Remove unused imports

jzelenjak committed:
Refactor parsing

Remove unnecessary comments, change exceptions to if-statements, change some variable names.

jzelenjak commented on sage.py:
```diff
@@ -1374,74 +1349,60 @@ def parse(unparsed_data, alert_labels=[], slim=False):
         sig = raw['alert']['signature']
         cat = raw['alert']['category']
-        
         severity = raw['alert']['severity']
```
`severity` variable does not seem to be used anywhere in this method. Was it left on purpose?

azqanadeem replied:
Ah, we do not use the default severity as Suricata defines it, we define our own based on attack stages. You can remove this line.

jzelenjak commented on sage.py:
```diff
@@ -1374,74 +1349,60 @@ def parse(unparsed_data, alert_labels=[], slim=False):
         __sev.add(severity)
 
-        
     '''_cats = [(id,c) for (id,c) in enumerate(__cats)]
```
This code block is also commented out and I am not sure what to do with that. The sets above (`__cats`, `__ips`, ...) are only used in this place.

jzelenjak committed:
Refactor removing duplicates

Fix plotting in `remove_duplicates` method, change some variable names for better readability, remove unused code blocks.

jzelenjak committed:
Fix comparison with MicroAttackStage.NON_MALICIOUS in

jzelenjak committed:
Merge branch 'main' into refactoring

jzelenjak committed:
Add missing signature files

jzelenjak committed:
Refactor loading data and plotting histogram

jzelenjak commented on sage.py:
```diff
@@ -1343,105 +95,78 @@ def load_IANA_mapping():
             continue
-            
+
         if not slim:
```
Is this `slim` part still needed? Same lines for "Attempted Information Leak" and "Not Suspicious Traffic" are present in the main part of the method. I am also not sure where `alert_labels` come from (they are not created anywhere, if I am not mistaken)

azqanadeem replied:
slim is set to false I think. Basically, the alert is parsed and then the get_attack_stage_mapping is called to get the attack stage that matches the alert signature. This is done for all cases

azqanadeem replied:
The badIP, "Not suspicious traffic", "attempted info leak" are all edge cases meant for the experimental datasets. A general user does not necessarily have to run this part of the code. For now I think we can keep it in, and leave a comment that this can be removed if it is not needed.

azqanadeem replied:
okay, so I see that alert_labels is always empty and because slim is set to false, the code around "Not suspicious traffic" and "attempted info leak" is basically unreachable. So I think the whole block (if slim:) can be removed.

jzelenjak replied:
Alright, I will remove the `if slim:` block (and the `slim` part)

jzelenjak commented on sage.py:
```diff
@@ -1588,108 +277,98 @@ def load_data(path, t):
     # Choice of: Suricata category usage or Micro attack stage usage?
     SURICATA_SUMMARY = False
-    cats = {'A Network Trojan was detected': 0, 'Generic Protocol Command Decode': 1, 'Attempted Denial of Service': 2, 'Attempted User Privilege Gain': 3, 'Misc activity': 4, 'Attempted Administrator Privilege Gain': 5, 'access to a potentially vulnerable web application': 6, 'Information Leak': 7, 'Web Application Attack': 8, 'Successful Administrator Privilege Gain': 9, 'Potential Corporate Pr ...
+    suricata_categories = {'A Network Trojan was detected': 0, 'Generic Protocol Command Decode': 1, 'Attempted Denial of Service': 2,
```
Is this `SURICATA_SUMMARY` still needed? You have mentioned in one of the previous comments that SAGE uses the micro attack stages (Moskal). I have tried running with `SURICATA_SUMMARY = True` and got an exception on the line below (`percentages[tid][suricata_categories[alert[6]]] += 1`: `KeyError: 'Decode of an RPC Query'`). Can it be the case that not all Suricata categories are present?

azqanadeem replied:
Yes suricata_summary is needed. It was meant to provide two configurations to plot the histograms, which I dont remember anymore, but the code is technically supposed to be functional.

azqanadeem replied:
And indeed, I think the issue is with the cats = {}. You may leave a comment next to where cats is defined that it needs to be updated

jzelenjak replied:
I will add a comment then

jzelenjak commented on sage.py:
```diff
@@ -1588,108 +277,98 @@ def load_data(path, t):
+            'Detection of a Network Scan': 11, 'Not Suspicious Traffic': 12, 'Potentially Bad Traffic': 13,
+            'Attempted Information Leak': 14}
 
     cols = ['b', 'r', 'g', 'c', 'm', 'y', 'k', 'olive', 'lightcoral', 'skyblue', 'mediumpurple', 'springgreen', 'chocolate', 'cadetblue', 'lavender']
```
This list is not used anywhere in the method. Can it be deleted?

jzelenjak committed:
Merge branch 'main' into refactoring

jzelenjak committed:
Refactor episode generation

jzelenjak commented on sage.py:
```diff
@@ -349,125 +279,132 @@ def load_data(path, t):
+    negative = [item for item in negative if item not in common]
+    positive = [item for item in positive if item not in common]
+
+    negative = [x for x in negative if (frequencies[x[0]] <= 0 or x[0] == len(frequencies) - 1)]
```
I was wondering what is the purpose of these two lines (originally, `negative = [x for x in negative if (y[x[0]] <= 0 or x[0] == len(y)-1)]` and `positive = [x for x in positive if (y[x[0]] <= 0 or x[0] == 0)]`, same functionality)? `y` or `frequencies` cannot be negative. I was looking at these lines, but could not really understand what they are for

azqanadeem replied:
positive has all positive slopes and negative has all negative slopes. Lines 385-385 (in your comment) are discarding slopes that are common between the two lists. `y' is a list with lengths of episodes. Lines 387-388 are special cases that keep the slope, i.e., for negative slope: if the episode is empty or it is the last item in the episode list, and for positive:  if the episode is empty, or it is the first item in the episode list. Essentially, these cases were added for sequences without a positive/negative pair.

jzelenjak replied:
I see. Thanks for the explanation. Would it be better to use `==` instead of `<=`? The `<` part was the one that confused me, since a length of an episode cannot be negative

jzelenjak commented on sage.py:
```diff
@@ -571,243 +485,159 @@ def getepisodes(action_seq, mcat, plot, debug=False):
+            # Alert format: (diff_dt, src_ip, src_port, dst_ip, dst_port, sig, cat, host, ts, mcat)
+            src_ip, dst_ip, signature, ts, mcat = alert[1], alert[3], alert[5], alert[8], alert[9]
+            dst_port = alert[4] if alert[4] is not None else 65000
+            # Simply respect the source,dst format! (Correction: source is always source and dest always dest!)
```
Here I also don't really understand. It says `Correction: source is always source and dest always dest!`, but there are still the if-checks for (src,dst) and (dst,src). Are they still needed?

azqanadeem replied:
Maybe this comment is outdated. Problem is that the check is dataset-dependent. For the datasets where we know attacker IPs, we can convert them into attacker,victim format (e.g., for cptc-2018), but for other datasets, we simply have to believe that src,dst == attacker,victim. I think make_condensed_data() is the function that's doing the real check based on IP format. You could, in theory, move that check in this function already (but make sure to check whether the dataset is cptc-201X, otherwise we get no AGs).

jzelenjak replied:
We can discuss it during the meeting, looking at a related comment later in the code. Also, there is no easy way to check for the dataset. We can check the directory name, but that can also be weirdly named (I think during RP I just had `cptc/2017`, but someone might also have `alerts/2017` if they are only using CPTC dataset

jzelenjak replied:
I have created an issue for this dataset part, you might check it and comment if necessary. I will resolve all other comments related to this issue

jzelenjak commented on sage.py:
```diff
@@ -571,243 +485,159 @@ def getepisodes(action_seq, mcat, plot, debug=False):
-            if((source,dest) in host_alerts.keys()):
-                host_alerts[(source,dest)].append((dest, c, ts, port, signature)) # TODO: remove the redundant host names
-                #print(source, dest, (micro[c].split('.'))[-1], port)
+            if (src_ip, dst_ip) not in host_alerts.keys() and (dst_ip, src_ip) not in host_alerts.keys():
```
These lines. Should they be kept? `(src_ip, dst_ip)` is added to `host_alerts` in either case

azqanadeem replied:
Yes. The first if() simply adds an empty list and creates the dict key. The second if,else() actually append the tuples in the respective list.

jzelenjak replied:
Yeah, I understood this, but I was a bit confused that it checks for both tuples, but adds only one

jzelenjak commented on sage.py:
```diff
@@ -571,243 +485,159 @@ def getepisodes(action_seq, mcat, plot, debug=False):
+
+                if len(raw_episodes) > 0:
+                    # For each episode, flatten relevant data from the windows of the corresponding alert sequence
+                    for epi in raw_episodes:
```
Maybe it is a bit difficult to understand exactly what has been changed here. Essentially, what I have done is flatten services, signatures etc. immediately and create an episode directly, instead of storing lists for services, signatures etc. and then flatten them for each episode in a list of episodes

UPDATE: this comment refers to the `_create_episode` function. I left the comment before I extracted this part, so this is why it is here

azqanadeem replied:
I cannot see this function in this file so I can't see what has been changed.

jzelenjak replied:
[Resolved during the meeting]

jzelenjak committed:
Refactor trace generation

jzelenjak commented on sage.py:
```diff
@@ -571,393 +485,271 @@ def getepisodes(action_seq, mcat, plot, debug=False):
-        team_episodes.append(t_ep)
+        team_episodes_.append(team_host_episodes)
         team_times.append(_team_times)
-    return (team_episodes, team_times)
```
`team_times` are not used anywhere else in the code. They are ignored when the method returns. Should I still keep them or can they be deleted?

azqanadeem replied:
Keep them. They store the first activation time for each team. Might be useful at some point.

jzelenjak commented on sage.py:
```diff
@@ -571,393 +485,271 @@ def getepisodes(action_seq, mcat, plot, debug=False):
     host_data = {}
-    #team_host_data= []
     print('# teams ', len(team_episodes))
     print('----- TEAMS -----')
```
Teams are printed twice in the logs: during episode generation and during subsequence generation. Is it intentional?

azqanadeem replied:
If you mean print(tid, sep=' ', end=' ', flush=True ) and print((tid+1), sep=' ', end=' ', flush=True), then I think the second one prints total number of subsequences, not teams. If that's true, we can rename the tid variable.

jzelenjak replied:
Not really. The first time it is called in `aggregate_into_episodes`, after alerts are grouped per team. The second time is in `host_episode_sequences` method when iterating over episodes groups by team.

![image](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/5a07ddb7-5228-40d6-b9c8-fea028a9cd66)
![image](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/e06f04e2-911d-465d-8fd5-3587e8a408c0)
![image](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/0bdefa90-7a4b-44de-8f39-d519ca036685)

jzelenjak commented on sage.py:
```diff
@@ -571,393 +485,271 @@ def getepisodes(action_seq, mcat, plot, debug=False):
+                subsequences.append(episodes)
+                continue
 
-            else:
```
Large block in the diff, but essentially I have simplified some nested if-elses

azqanadeem replied:
I expect you double checked that there are no missing edge cases?

jzelenjak replied:
There should not be any edge cases, since instead of nesting blocks in the "else" part (if `FULL_SEQ` is false or if `pieces < 1` is false), I have added `continue`
![image](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/77510f02-f211-4b20-99ca-88030909c570)
![image](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/88478d72-dd32-4b04-b698-bbdaee03d0ba)

jzelenjak replied:
As for the `if len(cuts) == 0:` part, the for loop `for j in range(len(cuts)):` will not execute, `subsequence` will be the same as `episodes` and `episodes < 2` have already been discarded above

![image](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/5cf7bb75-cb07-480d-a463-b4497ff2654a)
![image](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/5d8a1251-5868-4dc4-821f-00089631c730)

jzelenjak replied:
As a check: I wrote subsequences to a file both before and after, using CPTC-2017. The files are identical

jzelenjak commented on sage.py:
```diff
@@ -571,393 +485,271 @@ def getepisodes(action_seq, mcat, plot, debug=False):
+
+## 27 Aug 2020: Generating traces for FlexFringe
+def generate_traces(subsequences, datafile):
+    all_services = [[most_frequent(epi[6]) for epi in subseq] for subseq in subsequences]
```
What is the purpose of these lines? Is it to show all the unique services that occur in the AGs (so only mserv's should be counted), or to show all unique services present in all episodes?

In the latter case, `most_frequent` call will discard some episodes and the count of services will be lower than actually (which confused me a bit during RP). In the former case, the log string is misleading; also, these services can be counted in the for-loop below so that they are not recomputed twice.

azqanadeem replied:
It was used when we were grouping services by themes. Sometimes, new services would show up and since the themes dictionary was manually created, there would be errors. This can be removed now that we directly show the service itself.

jzelenjak replied:
Just to be sure, do you mean lines 655-658 in the screenshot below?

![image](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/4ee8500c-abd1-4468-8c10-00f29f81a407)

jzelenjak replied:
[Resolved during the meeting]

jzelenjak commented on sage.py:
```diff
@@ -571,393 +485,271 @@ def getepisodes(action_seq, mcat, plot, debug=False):
-        #print(stime)
-        #print(' '.join(mcats))
-
-        #multi = [str(c)+":"+str(n)+","+str(s) for (c,s,n) in zip(mcats,max_servs,num_servs)] # multivariate case
```
Is this multivariable case still relevant? I have tried running with this line and got `KeyError: '4'`  on the line `sev = rev_smallmapping[sym.split('|')[0]]` in `traverse` method. So either this line (and some unused variables above) can be removed if this is not relevant anymore, or it has to be fixed so that it runs (although, I cannot say what the fix will include other than `mcat` and how to check the results, since I don't have the 'original' version.

azqanadeem replied:
Yep, lets keep it in. The multivariate symbol is defined as: for each episode, the micro attack stage, the most frequently targeted service, the number of services in that episode.

azqanadeem replied:
And actually, this is why it is a bad idea to flatten the list of signatures/services etc. We want to preserve events on an episode basis.

jzelenjak replied:
Alright, however I am not sure how it should be fixed. Right now it is `[str(c) + ":" + str(n) + "," + str(s) for (c, s, n) in zip(mcats, max_services, num_services)]`, however all the splits in attack graphs and lower are based on `|` symbol. Furthermore, a frequently occurring pattern is `rev_smallmapping[sym.split('|')[0]])`, which would break in case of multi. Could it then be `mcat|mserv,num_serv` or something like that? I think services are not really checked

jzelenjak replied:
Also, I will update the variable to `symbols`, so that it can be used in both case (since `multi` is confusing if only `mcat|mserv` are used)

jzelenjak replied:
Actually, for the multivariable case, is the entire tuple a trace in FlexFringe or is it still `mcat|mserv`? Because if the former, then I am not sure if there will be any useful learning, and if the latter, then is it only needed for the states in the AGs?

![image](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/37b46dea-fbde-4f59-9ec8-89a439c5d4bf)
(This part is not doing anything for simply `mcat|mserv` but has to be modified depending on the format of the multivariable case

jzelenjak replied:
I have added a TODO, as discussed in the meeting

jzelenjak commented on sage.py:
```diff
@@ -571,393 +485,271 @@ def getepisodes(action_seq, mcat, plot, debug=False):
+
 ## 2 sept 2020: Learning the model
 def flexfringe(*args, **kwargs):
-  """Wrapper to call the flexfringe binary
```
This method was not indented with 4 whitespaces (this diff might be a bit confusing)

jzelenjak commented on sage.py:
```diff
@@ -349,125 +279,132 @@ def load_data(path, t):
+
+    for tid, team in enumerate(team_alerts):
+        for alert in team:
+            # if alert[9] == 999:
```
Is it still needed?

jzelenjak commented on sage.py:
```diff
@@ -349,125 +279,132 @@ def load_data(path, t):
             if SURICATA_SUMMARY:
-                #if cats[ev[6]] != 14:
-                    t[tid][cats[ev[6]]] += 1
+                # if suricata_cats[alert[6]] != 14:
```
Is this still needed?

jzelenjak committed:
Extract some parts into separate methods

jzelenjak commented on sage.py:
```diff
@@ -104,105 +97,78 @@ def load_IANA_mapping():
 def parse(unparsed_data, alert_labels=[], slim=False):
-    
+
     FILTER = False
```
I have kept all such capitalised parameters for now. They can be extracted as global parameters in another PR (if needed)

jzelenjak commented on sage.py:
```diff
@@ -104,105 +97,78 @@ def load_IANA_mapping():
-        __sev.add(severity)
 
-        
     '''_cats = [(id,c) for (id,c) in enumerate(__cats)]
```
Is this part still needed? If not, the sets above (`__cats`, `__ips` etc) can also be removed because they are not used anywhere

azqanadeem replied:
Indeed these sets can be removed. The initial idea was to do one hot encoding on them but it was not needed

jzelenjak committed:
Change some variable names and add some minor changes

jzelenjak committed:
Change the placement of two methods

jzelenjak commented on sage.py:
```diff
@@ -1,49 +1,34 @@
 
 
-def get_attack_stage_mapping(signature):
+def _get_attack_stage_mapping(signature):
```
I have prefixed the helper methods with an underscore (so that the main ones are easier to find and differentiate)

jzelenjak commented on sage.py:
```diff
@@ -84,14 +69,14 @@ def load_IANA_mapping():
     for row in table:
         # Drop missing port number, Unassigned and Reserved ports
-        if row[1] and 'Unassigned' not in row[3]:# and 'Reserved' not in row[3]:
+        if row[1] and 'Unassigned' not in row[3]:  # and 'Reserved' not in row[3]:
```
The "Reversed" that is commented out, should it be used or can it be removed?

UPDATE: I meant "Reserved", not "Reversed"

azqanadeem replied:
Reserved, not reversed. Keep it in, maybe IANA changes its mind again.

jzelenjak commented on sage.py:
```diff
@@ -104,105 +89,76 @@ def load_IANA_mapping():
 
-def parse(unparsed_data, alert_labels=[], slim=False):
-    
+# Step 1.1: Parse the input alerts
```
To make the codebase easier to follow, I have added steps

jzelenjak commented on sage.py:
```diff
@@ -104,105 +89,76 @@ def load_IANA_mapping():
-        #print(host)
-        ts = raw['timestamp']
-        dt = datetime.datetime.strptime(ts, '%Y-%m-%dT%H:%M:%S.%f%z')# 2018-11-03T23:16:09.148520+0000
-        DIFF = 0.0 if prev== -1 else round((dt - prev).total_seconds(),2)
```
PEP was complaining about capitalised local variables, so I have changed the name

UPDATE: this is not needed if these methods will move to another file

jzelenjak commented on sage.py:
```diff
@@ -215,259 +171,271 @@ def parse(unparsed_data, alert_labels=[], slim=False):
                 continue
             if spl[2] == 'Attempted Information Leak' and FILTER:
                 continue
-                
```
A bit confusing looking at this diff (because of the weird matching of blank lines)... Essentially, I have moved the plotting part from `remove_duplicates` method into a separate method. Btw, these plots look pretty nice :)

jzelenjak commented on sage.py:
```diff
@@ -215,259 +171,271 @@ def parse(unparsed_data, alert_labels=[], slim=False):
     plt.tight_layout()
-    #plt.show() 
-    return plt
+    plt.savefig('data_histogram-' + experiment_name + '.png')
```
Instead of returning `plt` and then calling `plt.savefig(...)`, I have moved this call into the method

jzelenjak commented on sage.py:
```diff
@@ -215,259 +171,271 @@ def parse(unparsed_data, alert_labels=[], slim=False):
+    # plt.show()
+
+
+def _group_alerts_per_team(_team_alerts):
```
The method `aggregate_into_episodes` was about 200 lines, so I have extracted some parts into separate methods (`_group_alerts_per_team`, `_create_episode`, and `_plot_alert_volume_per_episode`)

jzelenjak commented on sage.py:
```diff
@@ -215,259 +171,271 @@ def parse(unparsed_data, alert_labels=[], slim=False):
+            # Say 'unknown' if the port cannot be resolved
+            dst_port = 'unknown' if (dst_port not in port_services.keys() or port_services[dst_port] == 'unknown') else port_services[dst_port]['name']
+
+            if (src_ip, dst_ip) not in host_alerts.keys() and (dst_ip, src_ip) not in host_alerts.keys():
```
Is it necessary to check for the reversed (src_ip, dst_ip)? There was already a similar part with doing inverse of the attacker somewhere below

jzelenjak replied:
[Resolved during the meeting]

jzelenjak commented on sage.py:
```diff
@@ -215,259 +171,271 @@ def parse(unparsed_data, alert_labels=[], slim=False):
+    return team_data
+
+
+def _get_ups_and_downs(frequencies, slopes):
```
I have extracted some parts from `get_episodes` method into `_get_ups_and_downs` (which was also mentioned in the paper) and `_plot_episodes`

azqanadeem replied:
Good thinking following the paper convention, thanks!

jzelenjak commented on sage.py:
```diff
@@ -215,259 +171,271 @@ def parse(unparsed_data, alert_labels=[], slim=False):
+
 
-## 14
-def legend_without_duplicate_labels(ax, fontsize=10, loc='upper right'):
```
Moved this function lower (above the function `_plot_alert_volume_per_episode` where it is called)

jzelenjak commented on sage.py:
```diff
@@ -481,483 +449,311 @@ def getepisodes(action_seq, mcat, plot, debug=False):
+    end_time = (last_ts - start_times[tid]).total_seconds()
+    period = end_time - start_time
+
+    episode = (start_time, end_time, mcat, len(events), alert_volume, period, services, unique_signatures, (first_ts, last_ts))
```
In the episode definition, there was `len(raw_events)`, however in the code there was just `events` (or `x[3]`). The value is not used anywhere else, but should it actually be `len(raw_events)`?

azqanadeem replied:
No, events is good enough. Len() can be computed later too, as needed, but the events cant be extracted if only len() is stored.

jzelenjak replied:
Right, the `events` list is a list of lengths of the alerts windows in the alert sequence for this episode, so it is essentially the length, and this is probably what was meant in the episode definition. My bad

jzelenjak commented on sage.py:
```diff
@@ -481,483 +449,311 @@ def getepisodes(action_seq, mcat, plot, debug=False):
-    #print(len(s_t))
-    team_episodes = []
 
+def _create_episode(alert_seq_epi, mcat, tid):
```
This function is essentially a compact version of all the long list comprehensions from the `aggregate_into_episodes` function

azqanadeem replied:
I hope you kept the commented out test cases still in the code.

jzelenjak replied:
Yes, but they are in `_get_episodes` method. I have added a TODO that they could be moved to a separate test file.

jzelenjak commented on sage.py:
```diff
@@ -481,483 +449,311 @@ def getepisodes(action_seq, mcat, plot, debug=False):
-                
-                episodes = getepisodes(mindata, micro[mcat], False)
+            # Alert format: (dst_ip, mcat, ts, dst_port, signature)
+            # print(attacker_victim, len([(x[1]) for x in alerts])) # TODO: what about IPs that are not attacker related?
```
Is the TODO still relevant?

azqanadeem replied:
lol yes

jzelenjak commented on sage.py:
```diff
@@ -481,483 +449,311 @@ def getepisodes(action_seq, mcat, plot, debug=False):
+                alert_seq = []
+                for i in range(int(first_elapsed_time), int(relative_elapsed_time[-1]), step):
+                    window = [a for dt, a in zip(relative_elapsed_time, alerts) if (i <= dt < (i + step)) and a[1] == mcat]
+                    alert_seq.append(window)  # Alerts per 'step' seconds (window)
```
From what I have understood, this is essentially the alert sequence that was mentioned in the paper, so I have renamed it accordingly (I don't know what the `mindata` referred to)

azqanadeem replied:
It was short for minute_data, i.e., how many alerts within a single 2.5-minute window were triggered. Perhaps a better name for this would be hyperalert_seq (since each element captures multiple alerts)

jzelenjak replied:
Hmm, that's if fine, but in the paper, `alert sequence` was explicitly defined as a windowed list, which is why I used this name (for consistency). Do you want me to rename it to `hyperalert_seq`?

![image](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/d27c8a1d-92b9-4132-8656-5d628904fa27)

jzelenjak replied:
Updated, as discussed in the meeting

jzelenjak commented on sage.py:
```diff
@@ -481,483 +449,311 @@ def getepisodes(action_seq, mcat, plot, debug=False):
+
+
+# Step 2: Create alert sequence and get episodes
+def aggregate_into_episodes(_team_alerts, step=150):
```
PEP was often complaining about local variables/parameters shadowing the global ones (which I pretty much agree with), so I have prefixed these local variables/parameters with an underscore

UPDATE: this is not needed if these methods move to another file

jzelenjak commented on sage.py:
```diff
@@ -481,483 +449,311 @@ def getepisodes(action_seq, mcat, plot, debug=False):
-            #print(att)
-            if att not in host_data.keys():
-                host_data[att] = []  
-            #EPISODE DEF: (startTime, endTime, mcat, len(rawevents), volume(alerts), epiPeriod, epiServices, list of unique signatures)
```
After the changes, this episode definition in the comment is not needed, since `_create_episode` contains this line but for actually creating the episode tuple

azqanadeem replied:
Could you then add a comment in text (next to episode creation) describing what is actually in an episode?

jzelenjak replied:
Alright, I have added it above the line in `_create_episode` method. After I have renamed the variables, they are essentially the same as in the comment, that's why I did not keep it initially

jzelenjak commented on sage.py:
```diff
@@ -481,483 +449,311 @@ def getepisodes(action_seq, mcat, plot, debug=False):
+            _subsequences[attacker_victim + '-' + str(count)] = subsequence
+
+    print('\n# sub-sequences:', len(_subsequences))
+    return _subsequences
```
Instead of `alerts` and `keys` (which are not obvious what they refer to), I have made it one dictionary `subsequences`. They are recombined (zipped) after the S-PDFA learning anyway, also it simplifies this method a bit (compared to adding separately to keys and separately to alerts)

jzelenjak commented on sage.py:
```diff
@@ -481,483 +449,311 @@ def getepisodes(action_seq, mcat, plot, debug=False):
+def generate_traces(subsequences, datafile):
+    all_services = [[_most_frequent(epi[6]) for epi in subseq] for subseq in subsequences.values()]
+    print('----- All unique services -----')
+    print(set([service for services_subseq in all_services for service in services_subseq]))
```
What exactly is the purpose of this part? To compute all services present in the episodes or only the ones present in the AGs (i.e. mserv's)? In the former case, I don't get the use of `most_frequent`. In the latter case, the log is a bit misleading, also this can be computed in the for-loop below where mserv's are actually computed

jzelenjak commented on sage.py:
```diff
@@ -1629,126 +1425,111 @@ def make_AG(condensed_v_data, condensed_data, state_groups, sev_sinks, datafile,
-t = float(sys.argv[3])
-w = int(sys.argv[4])
-_s_,_e_ = 0, 100
+path_to_json_files = sys.argv[1]
```
I have renamed these variables to the ones defined in "Usage", since they are easier to understand

jzelenjak commented on sage.py:
```diff
@@ -1629,126 +1425,111 @@ def make_AG(condensed_v_data, condensed_data, state_groups, sev_sinks, datafile,
+# ----- MAIN ------
 
 if len(sys.argv) < 5:
-    print('USAGE: sage.py {path/to/json/files} {experiment_name} {alert_filtering_window (def=1.0)} {alert_aggr_window (def=150)} {(start_hour,end_hour)[Optional]}')
```
Usually, Unix man-pages have "Usage", that's why I have changed it

azqanadeem replied:
A better idea: replacing this manually written code with ConfigParser (see example in SECLEDS). Makes it easier to handle the command line params.

jzelenjak replied:
Do you mean `ArgumentParser`? I think `ConfigParser` is for an init file (in either case, this can be left for another PR)

![image](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/1002e6ea-67fa-4dcc-b136-b92a17512b61)

jzelenjak replied:
I have created an issue for this part, you might check it and comment if necessary. I will resolve all other comments related to this issue

jzelenjak commented on sage.py:
```diff
@@ -1629,126 +1425,111 @@ def make_AG(condensed_v_data, condensed_data, state_groups, sev_sinks, datafile,
+print('------ Learning S-PDFA ---------')
 flexfringe(path_to_traces, ini=path_to_ini, symbol_count="2", state_count="4")
 
-## Copying files
```
In this part, there were multiple variables referring to the same thing (`path_to_model`, `path_to_traces`, `datafile`, `modelname`, `outfile`). I have kept only `path_to_traces` to avoid the confusion

jzelenjak commented on sage.py:
```diff
@@ -481,483 +449,311 @@ def getepisodes(action_seq, mcat, plot, debug=False):
+            # Cut episode sequence when a low-severity episode follows a high-severity episode
+            count = 0
+            mcats = [epi[2] for epi in episodes]
+            cuts = [i for i in range(len(episodes) - 1) if (len(str(mcats[i])) > len(str(mcats[i + 1])))]  # (ep[i] > 100 and ep[i+1] < 10)]
```
This is the part I would like to discuss in more detail. During RP, I have noticed that there were many traces from the root node of the S-PDFA that were low-severity episodes, which means that episode subsequence ended with a low-severity state. This, however, should not happen because SAGE accentuates high-severity states (one of its strong points).

Firstly, looking at this line, `len(str(mcats[i])) > len(str(mcats[i + 1]))` will also be true for med-sev episode and then low-sev episode (e.g. from FF trace file: "ACE|microsoft-ds ACE|microsoft-ds exfil|microsoft-ds", ACE is med-sev and it is at the end of the episode subsequence). So, I think the version that is currently commented out is closer to the truth (or maybe even `len(str(mcats[i])) == 3 and len(str(mcats[i])) > len(str(mcats[i + 1]))`, i.e. a low- or mid-sev episode follows a high-sev episode).

Secondly, it might be the case that there is no such cut possible (e.g. the entire episode sequence consists of low- and med-sev episodes), or that the part after the last cut consists only of low- and medium severity episodes (meaning that this last part is not an episode subsequence, since it does not end with a high-severity episode). The question then arises whether such parts should be kept or whether they can be discarded.

We can also discuss it later (maybe even during a meeting), since this is not part of the refactoring and should be handled in a different PR. I am just writing it here so that I don't forget this.

azqanadeem replied:
Good point! So if we look at the last alert distribution for the alert subsequences, we see that most of them end with low-sev alerts. It's likely that most attackers do probing frequently between other attack attempts. Many traces also only ever end in med-severity events. This is why of course high-sev events are rare. 

Anyhow, this part of the code only needs to cut the sequences at appropriate places -- no discarding is needed (the discarding already happens later in make_AG()). So here, the viable cuts are: [med, low], [high, low], and [high, med]. For the latter two cases, we will see the first subsequence in the AGs related to the objective = high. For the first case, unless AGs are also made for medium severity objectives, they will not be appearing anywhere (and we call them partial paths).

jzelenjak replied:
I see. I think it might be documented somewhere. Since I got confused, as the paper had this line:
![image](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/ca562cfc-9337-48af-a924-7d75325a66d4)

azqanadeem replied:
The proposed solution for this one was to add a comment where subsequences are created regarding the 3 valid cut criteria.

jzelenjak committed:
Refactor loading model and encoding sequences

jzelenjak committed:
Refactor creating state sequences

jzelenjak committed:
Refactor state groups and grouping episodes

jzelenjak committed:
Refactor part of  method

jzelenjak committed:
Refactor another part of  function

jzelenjak committed:
Refactor  method

jzelenjak committed:
Fix missing sinks

jzelenjak committed:
Change some parts

jzelenjak commented on sage.py:
```diff
@@ -1,49 +1,34 @@
 IANA_CSV_FILE = "https://www.iana.org/assignments/service-names-port-numbers/service-names-port-numbers.csv"
-IANA_NUM_RETRIES = 3
+IANA_NUM_RETRIES = 5
 DB_PATH = "./ports.json"
```
I would assume, this variable is outdated?

azqanadeem replied:
If you mean DB_PATH, then yes.

jzelenjak committed:
Revert DOCKER back to True

I am sorry, this time I forgot to change it back :(

jzelenjak commented on sage.py:
```diff
@@ -104,105 +89,72 @@ def load_IANA_mapping():
+        dst_port = None if 'dest_port' not in raw.keys() else raw['dest_port']
+
+        # Filter out mistaken alerts / uninteresting alerts
+        if src_ip == badIP or dst_ip == badIP or cat == 'Not Suspicious Traffic':  # TODO: add `filter_alerts` here?
```
As already discussed, I removed the "slim" part. Do you want to add boolean`filter` here? This might change the resulting AGs though

azqanadeem replied:
yes, lets add the filter. The final point of this whole refactoring exercise is to have identical attack graphs as before. So set the flag values appropriately please.

jzelenjak replied:
I have only written it in the TODO, the flags are the same as they were before

jzelenjak replied:
I have created an issue for this part, you might check it and comment if necessary. I will resolve all other comments related to this issue

jzelenjak commented on sage.py:
```diff
@@ -481,1274 +423,920 @@ def getepisodes(action_seq, mcat, plot, debug=False):
-            vic = attacker[1]
-            #print(perp)
-            #if ('10.0.0' in perp or '10.0.1' in perp):
+            # if ('10.0.0' in attacker or '10.0.1' in attacker):
```
I guess, this is for one of the three datasets? Leaving like that?

azqanadeem replied:
Yes, testing for specific cases

jzelenjak commented on sage.py:
```diff
@@ -481,1274 +423,920 @@ def getepisodes(action_seq, mcat, plot, debug=False):
+        dfa[edge["source"]][edge["name"]] = edge["target"]
+
+    # Even though the properties might not be needed, the node has to be present in the dictionary (for sinks)
+    for entry in machine["nodes"]:
```
So, as the comment says...

azqanadeem replied:
We need to check with Sicco for this one.

jzelenjak replied:
I think the point here was that as soon as there is a transition from a non-sink node to a sink node, then this sink has to be defined in the `sinks` variable (in the `_traverse` method). I once removed the "nodes" part here, but got some states in AGs that used to be sinks but were not anymore. So I have kept this part here, just that I have added the properties that might be useful (at least they were useful for me during RP, even though I checked them in the JSON files from FlexFringe)

jzelenjak commented on sage.py:
```diff
@@ -481,1274 +423,920 @@ def getepisodes(action_seq, mcat, plot, debug=False):
+        state = dfa[state][sym]
+        if state == "":
+            # Only keep IDs of medium- and high-severity sinks
+            if len(str(sev)) >= 2:
```
This part here is not really correct. It sort of worked, not really in an intended way. As I have found out after quite some debugging (since I wrote an if-check here, as mentioned in the comment), defaultdict's in Python create a new element when it is not present in the dict and is accessed (which was the case here). This leads to `-1` being added to (global) `sinks` variable, which should not be there (otherwise states with ID -1 will sort of be sinks).

This was the issue for CCDC dataset. In the image you can see a (still reversed) sequence where `netDOS` follows `vulnD` (which should not be the case, see also my comment above about splitting subsequences, but let's assume it is). Because the ID is removed from `vuldD` (as it is low-sev node), the next transition is from state `-1` which of course cannot be correct as there are no nodes in S-PDFA with ID -1 (except for one dummy node, but that's not a problem here).

![issue-non-sev-sinks](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/13e84305-00fe-4fdf-9226-5a63571db8d4)

The states are essentially the same, except for this -1 ID on some states. As for solutions, the IDs can be stored in a separate list here (only for transitions, while the one with `-1`s can be returned) or ID can be removed later. In either case, I can handle it in a separate PR (depending on your reply), since it will change the resulting AGs and I will not have a baseline to compare to for this PR

azqanadeem replied:
Let's discuss this in a meeting. I dont really understand why we have a subsequence that starts in a high event.

azqanadeem replied:
Create an issue for this. Might be a potential bug.

jzelenjak replied:
Actually, now that I think of it, I remember Senne showed a graph that started with a high-severity node...
![image](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/4f76d5a7-e30e-4fe5-98cd-3efd9f19e792)

jzelenjak replied:
I have created an issue for this (with all the images): #28

jzelenjak replied:
I have also created an issue for the sinks: #30

jzelenjak commented on sage.py:
```diff
@@ -481,1274 +423,920 @@ def getepisodes(action_seq, mcat, plot, debug=False):
+                    state = sinks[state_list[-1]][sym]
+                    state = '-1' if state == '' else state
+                except IndexError:
+                    state = '-1'  # With `printblue = 1` in spdfa-config.ini this should not happen
```
...and after what I have written in the comment above, though this part can still be kept

jzelenjak commented on sage.py:
```diff
@@ -481,1274 +423,920 @@ def getepisodes(action_seq, mcat, plot, debug=False):
-    #print(len(traces), len(state_traces))
-    print('Traces in sinks: ', num_sink, 'Total traces: ', total, 'Percentage: ',100*(num_sink/float(total)))
-    return (traces, state_traces)    
+        total_traces += len(state_list)  # TODO: decide what to do with the root
```
Right now, the root node (`0`) is in `state_traces`. I think it should not be counted as trace, since it is artificial, but what do you think?

azqanadeem replied:
The length of a subsequence is one less than the length of a state sequence (since [0] is added for an artificial transition). The new_state list actually gets rid of [0]. Maybe that format should be followed,

jzelenjak replied:
But that happens later (in `make_condensed_data` or `make_state_sequences`), so right now the root state is still counted from what I can see. Should I then do the cut `s[1:]` when the arrow is? 
![image](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/5ad57382-4d93-4e60-8018-2d328d55338d)

jzelenjak replied:
[Updated as discussed in the meeting]

jzelenjak commented on sage.py:
```diff
@@ -481,1274 +423,920 @@ def getepisodes(action_seq, mcat, plot, debug=False):
-    print('Traces in sinks: ', num_sink, 'Total traces: ', total, 'Percentage: ',100*(num_sink/float(total)))
-    return (traces, state_traces)    
+        total_traces += len(state_list)  # TODO: decide what to do with the root
+        traces_in_sinks += state_list.count('-1')  # Add low-severity sinks (i.e. stateID = -1)
```
When fixing the part I have mentioned above, nodes with ID -1 are essentially low-severity sinks, so I think they should also be counted

azqanadeem replied:
In the original code, everything in sinks was being considered, right? even low-sev sinks.

jzelenjak replied:
Yes, you are right (I have updated this place as it was)

jzelenjak commented on sage.py:
```diff
@@ -481,1274 +423,920 @@ def getepisodes(action_seq, mcat, plot, debug=False):
+        traces_in_sinks += state_list.count('-1')  # Add low-severity sinks (i.e. stateID = -1)
+        traces_in_sinks += len(_sev_sinks)         # Add medium- and high-severity sinks (i.e. stateID != -1)
 
-def find_severe_states(traces, m, m2):
```
I have merged `encode_sequences` and `find_severe_states` into one method, as there were no obvious reason to keep them separate (`traverse` method was called twice)

azqanadeem replied:
That's fine. Anyway the find_severe_states() is not being really used anywhere if I remember correctly.

jzelenjak commented on sage.py:
```diff
@@ -481,1274 +423,920 @@ def getepisodes(action_seq, mcat, plot, debug=False):
+                count += 1
+            subsequence = episodes[rest[0]:rest[1]+1]
+            if len(subsequence) < 2:
+                # print('discarding symbol ', [x[2] for x in al]) # TODO This one is not cool1
```
Is the TODO still relevant?

azqanadeem replied:
Yep. There is a problem in this code: when splitting, (if I remember correctly) if a sequence goes like [low low medium high low], then [low low medium high] is saved but the last [low] is just discarded. We probably shouldnt lose alerts like this. But idk what to do with a single event either. Maybe we keep them regardless?

jzelenjak replied:
Hmm, for medium and high - maybe, for low - not sure. But for this one episode here, I don't think it is really useful. If AGs accentuate high-severity objectives, then this `high` in your sequence would be the end of an attack attempt. The next one (low in your case) does not give that much information. Also, wouldn't it mean giving a trace to FlexFringe with one symbol? I am not sure if FlexFringe will be happy to receive such one symbol, since it doesn't even have a transition

jzelenjak replied:
I have created an issue for this part: #29

jzelenjak commented on sage.py:
```diff
@@ -481,1274 +423,920 @@ def getepisodes(action_seq, mcat, plot, debug=False):
+        trace = [int(state) for state in state_traces[counter]]
+        max_services = [_most_frequent(epi[6]) for epi in episode_subsequence]
+
+        if 0 in trace and (not set(trace).isdisjoint(sev_states) or not set(trace).isdisjoint(med_states)):
```
I don't really understand the `level_one` here. All traces (unless `0` is removed above) start with `0` (the root node). And the if-check: if this trace has med- or high-sev nodes? Below, the print statement says "High-severity objective states", but is it the case? And those objective states are calculated when creating AGs...

azqanadeem replied:
This part can be removed. Probably was being used for testing at some point.

jzelenjak commented on sage.py:
```diff
@@ -481,1274 +423,920 @@ def getepisodes(action_seq, mcat, plot, debug=False):
         counter += 1
         
-        if '10.0.254' not in attacker:
+        if '10.0.254' not in attack:  # TODO: this part has to be commented for CCDC dataset
```
These checks for CCDC, I keep them for now, but I feel like commenting them out before running CCDC (or maybe even other datasets) is not the best option for the users...

azqanadeem replied:
Probably need to add a flag that if set=true for ccdc just runs relevant code blocks.

jzelenjak replied:
We can discuss it, since this pattern with datasets appears in multiple places and, as I wrote above, is not that easy to check (unless there is a parameter there, as you said)

jzelenjak commented on sage.py:
```diff
@@ -481,1274 +423,920 @@ def getepisodes(action_seq, mcat, plot, debug=False):
+        attacker_victim = team + '-' + attacker + '->' + victim
+        attacker_victim_inv = team + '-' + victim + '->' + attacker
+        inv = False
+        if '10.0.254' in victim:
```
I think this is CPTC-specific, but I am also not sure what to do with this, so for this PR it can be here. I think the reversing happens only here, i.e. after FlexFringe...

azqanadeem replied:
yeah, we probably need to do this as soon as the alerts are parsed and then keep the order. This comes way too late

jzelenjak commented on sage.py:
```diff
@@ -481,1274 +423,920 @@ def getepisodes(action_seq, mcat, plot, debug=False):
+
+
+# Step 6.3: Create state sequences (collecting sub-behaviors back into the same trace, augmented with state IDs)
+def make_state_sequences(episode_subsequences, state_traces, med_states, sev_states):
```
`condenced_data` is essentially state sequences, from what I can see in the paper

azqanadeem replied:
Yes

jzelenjak commented on sage.py:
```diff
@@ -481,1274 +423,920 @@ def getepisodes(action_seq, mcat, plot, debug=False):
+                        written.append(c)
+                        state_groups['ACTIVE_RECON'].add(node)
+                    print('ERROR: manually handled', node, ' in ACTIVE_RECON')  # TODO: include edges or not?
+            '''edge_line = re.match('\\D+(\\d+)\\s->\\s(\\d+)\\s\\[label=.*', line)  # 0 -> 1 [label=
```
I have kept the part for the edges commented. I have tried running both versions, and I feel like the one without edges is more correct (which is difficult to see since the labels are not really visible due to the size of the S-PDFA). On the other hand, there might be nodes that are only defined in edges. So, I have added a TODO for later...

jzelenjak commented on sage.py:
```diff
@@ -481,1274 +423,920 @@ def getepisodes(action_seq, mcat, plot, debug=False):
+    victim_episodes = {k: v for k, v in sorted(victim_episodes.items(), key=lambda kv: len([epi[0] for epi in kv[1]]))}
+    print('Victims hosts: ', set([team_victim.split('-')[-1] for team_victim in victim_episodes.keys()]))
+
+    attacker_episodes = dict()  # Episodes per (team, attacker)
```
Attacker data is not used anywhere. Still keeping it?

azqanadeem replied:
Not used but keep it still

jzelenjak commented on sage.py:
```diff
@@ -481,1274 +423,920 @@ def getepisodes(action_seq, mcat, plot, debug=False):
     return new_label
-    
-## Per-objective attack graph for dot: 14 Nov (final attack graph) 
-def make_AG(condensed_v_data, condensed_data, state_groups, sev_sinks, datafile, expname):  
```
I have extracted some parts into separate methods, as the original one was really huge

jzelenjak commented on sage.py:
```diff
@@ -481,1274 +423,920 @@ def getepisodes(action_seq, mcat, plot, debug=False):
+    cat = micro[episode[2]].split('.')[1]
+    signs = episode[5]
+    timestamps = episode[6]
+    if episode[3] in in_main_model:  # TODO: this check is useless, since it is always True
```
Here is another discussion. `in_main_model` seems useless, since it is created from the state IDs from the episodes and then the same states are queried in `in_main_model` (so all states will be in `in_main_model`, including the ones with ID `-1`). So, all the if-checks for `|Sinks` are never true and are currently not used. What could be checked instead is `if episode[3] != -1`, since nodes with ID -1 are low-severity sinks (I think, but also see the comment above in the traverse method). This will lead to some states having `|Sink` ID (which I saw on the image in some of the papers, but which is not the case for the current implementation), and as a result a bit more states (these states will now have an ID "Sink" and hence will not be merged with a non-sink nodes without ID (e.g. some VulnD).

I am not sure which solution you would prefer, but this has to be handled in a separate PR (since it changes the resulting AGs). If something from what I have written is unclear, we can also discuss it later

azqanadeem replied:
Lets discuss this. Ideally, I want all sinks states to have |Sink, even the high-sev ones that we salvaged. I think this is the only solution to the context problem (where multiple contextually different objectives are shown but some are directly from sinks so no learning was applied).

jzelenjak replied:
Hmm, not sure about this. If there are two mid-sev sinks and they have ID `Sink`, then graphviz will merge them as identical nodes, *no matter where they appear in the attack path and what is their context*. So, yes, this can be discussed

azqanadeem replied:
Created #27 for it

jzelenjak replied:
Thank you! I just wanted to make sure: is it an issue "for later"? In the sense that, do I still replace `if episode[3] in_main_model` with `if episode[3] != -1` (in a separate issue, since it will change the AGs)?

azqanadeem replied:
The code can stay like it is for now. The issue we created can deal with it later.

jzelenjak commented on sage.py:
```diff
@@ -481,1274 +423,920 @@ def getepisodes(action_seq, mcat, plot, debug=False):
+        if team_attacker not in team_attack_attempts.keys():
+            team_attack_attempts[team_attacker] = []
+        team_attack_attempts[team_attacker].extend(attempts)
+        # team_attack_attempts[team_attacker] = sorted(team_attack_attempts[team_attacker], key=lambda item: item[1])
```
Not sure if it needs to be sorted, so keeping it commented as it was

azqanadeem replied:
idk either

jzelenjak commented on sage.py:
```diff
@@ -481,1274 +423,920 @@ def getepisodes(action_seq, mcat, plot, debug=False):
+    return team_attack_attempts, observed_obj
+
+
+def _print_unique_attempts(team_attacker, attempts):
```
Extracted this

jzelenjak commented on sage.py:
```diff
@@ -481,1274 +423,920 @@ def getepisodes(action_seq, mcat, plot, debug=False):
-                        break
-                if sinkflag:
-                    lines.append((0,'"'+translate(obj)+'" [style="filled,dotted", fillcolor= salmon]'))
+            print('\t!!!! Objective', objective)
```
I have moved the print for Objective lower, so that it is only printed if the AG is actually created. This makes it easier to see from the logs for each victim which AGs have been created and for which victim no AGs have been created

azqanadeem replied:
Maybe keep this line as it is, and once the graph is actually created, we print "created". Then, at least we know the code is checking these objective options even if it doesnt save them.

jzelenjak replied:
Something like this, or do you also want to print if it is *not* created?
![image](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/66218bef-2e8e-4aeb-b24f-ccc8c168d55d)

jzelenjak replied:
[Updated as discussed in the meeting]

jzelenjak commented on sage.py:
```diff
@@ -481,1274 +423,920 @@ def getepisodes(action_seq, mcat, plot, debug=False):
+                if _is_severe_sink(obj_variant, sev_sinks):
+                    lines.append('"' + _translate(obj_variant) + '" [style="filled,dotted", fillcolor= salmon]')
                 else:
-                    lines.append((0,'"'+translate(obj)+'" [style=filled, fillcolor= salmon]'))
```
These 0's in the `lines` were not used anywhere and I haven't thought of any ways they could be used

azqanadeem replied:
This is explicitly for the generation of the .dot file. 0 makes sure these lines are at the top when we sort the list.

jzelenjak replied:
Oh, but the list with lines is not sorted anywhere. In this case, do you me still add these 0's back **and** sort the list? Because currently AGs are generated fine

azqanadeem replied:
Revert this change. We keep the tuple format anyway.

jzelenjak replied:
Done

jzelenjak commented on sage.py:
```diff
@@ -481,1274 +423,920 @@ def getepisodes(action_seq, mcat, plot, debug=False):
-                ones = [''.join([action[0] for action in attempt]) for attempt in attempts]
-                unique = len(set(ones)) # count exactly unique attempts
-                #print(unique)
-                #print('team', attackerID, 'total paths', len(attempts), 'unique paths', unique, 'longest path:', max([len(x) for x in attempts]), \
```
I think this part was used in the main paper, so I have kept it, but I have extracted it in a separate method, so that it is more readable and only one line has to be commented out

jzelenjak commented on sage.py:
```diff
@@ -481,1274 +423,920 @@ def getepisodes(action_seq, mcat, plot, debug=False):
-                lines.append((1, '"'+translate(vname)+'"'+' [tooltip="'+ "\n".join(signatures) +'"]'))
-            lines.append((1000,'}'))
-            
-            for l in lines: # count vertices and edges
```
I think this part was used in the main paper, so I have kept it, but I have extracted it in a separate method, so that it is more readable and only one line has to be commented out

jzelenjak commented on sage.py:
```diff
@@ -481,1274 +423,920 @@ def getepisodes(action_seq, mcat, plot, debug=False):
+                        vname = action[0]
+                        if vid == 0:  # If first action
+                            if 'Sink' in vname:  # If sink, make dotted TODO: this check is always False
+                                lines.append('"' + _translate(vname) + '" [style="dotted,filled", fillcolor= yellow]')
```
This is not executed (see comment above)

azqanadeem replied:
How come?

azqanadeem replied:
If vid is 0, that means it is the first action in a path. The code checks whether it is a sink or not. based on that, it sets the style of the vertex in the graph. Then comes the else condition (which is true for all the other actions in the path). Same process: checks if the action is in Sink, then set the style appropriately. Are you sure your indenting is correct?

jzelenjak replied:
I only meant that `if 'Sink' in vname:` block is not executed, since as I have mentioned above, there are no states with 'Sink' in their name.

jzelenjak commented on sage.py:
```diff
@@ -481,1274 +423,920 @@ def getepisodes(action_seq, mcat, plot, debug=False):
+                                else:  # Else, normal starting node
+                                    lines.append('"' + _translate(vname) + '" [style=filled, fillcolor= yellow]')
+                        else:  # For other actions
+                            if 'Sink' in vname:  # TODO: this check is always false
```
This is not executed (see comment above). Also, in this case, the `else: # For other actions` block is not doing anything, because of the if-check for`Sink`

jzelenjak replied:
For a better view, this is the original code on `main` branch:

![image](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/9dddd826-2ec9-4dfc-8f88-6805c1ab6410)

The `if 'Sink' in vname:` part is always false, as discussed above

jzelenjak commented on sage.py:
```diff
@@ -481,1274 +423,920 @@ def getepisodes(action_seq, mcat, plot, debug=False):
-                                    lines.append((0,partial+'"]'))
-
-                    # transitions
-                    bi = zip(attempt, attempt[1:]) # make bigrams (sliding window of 2)
```
I have also extracted this part into a separate method for readability. For now, I have kept all the indentations as originally (which was not always consistent), so that I can compare the AGs with the version before the PR

jzelenjak commented on sage.py:
```diff
@@ -481,1274 +423,920 @@ def getepisodes(action_seq, mcat, plot, debug=False):
+
+print('------ Reading alerts ------')
+(team_alerts, team_labels) = load_data(path_to_json_files, alert_filtering_window)
+# plot_histogram(team_alerts, team_labels)
```
I have commented this out, as it is not really needed for AGs and can always be uncommented if needed

azqanadeem replied:
No, keep the plot_histogram() in. The plot should always be generated and serves as a quick check for whether the alerts were correctly parsed.

jzelenjak replied:
Alright. Just that I remember that during RP (and also now) I was deleting these plots after every run and almost never looked at them. These plot was used in the main paper, so I thought it is only needed for this purpose

jzelenjak commented on sage.py:
```diff
@@ -481,1274 +423,920 @@ def getepisodes(action_seq, mcat, plot, debug=False):
-(med_states, sev_states, sev_sinks) = find_severe_states(traces, m, m2)    
-condensed_data = make_condensed_data(alerts, keys, state_traces, med_states, sev_states)
+# print('------ Clustering state groups ------')
+# state_groups = make_state_groups(state_sequences, path_to_traces)
```
I have commented this out, as it is not really needed for AGs and can always be uncommented if needed

azqanadeem commented on sage.py:
```diff
@@ -481,88 +418,65 @@ def getepisodes(action_seq, mcat, plot, debug=False):
-        plt.ylabel('Slope')
-        plt.plot(y, 'gray')
+        plt.ylabel('Frequency')
+        plt.plot(frequencies, 'gray')
```
Can you rename the y-axis back to slope? That is more representative of what we are showing in the plot.

jzelenjak committed:
Address some comments on the Pull Request

jzelenjak committed:
Address remaining comments on the Pull Request
